qiskit/transpiler/passes/crosstalk_adaptive_schedule.py

```diff
@@ -0,0 +1,717 @@
+                gate on qubit 2.
+            weight_factor (float): weight of gate error/crosstalk terms in the objective
+                weight_factor*fidelities + (1-weight_factor)*decoherence errors.
+                Weight can be varied from 0 to 1, with 0 meaning that only decoherence
+                errors are optimized and 1 meaning that only crosstalk errors are optimized.
+                weight_factor should be tuned per application to get the best results.
+            measured_qubits (list): a list of qubits that will be measured in a particular circuit.
+                This arg need not be specified for circuits which already include measure gates.
+                The arg is useful when a subsequent module such as state_tomography_circuits
+                inserts the measure gates. If CrosstalkAdaptiveSchedule is made aware of those
+                measurements, it is included in the optimization.
+        Raises:
+            ImportError: if unable to import z3 solver
+
+        """
+        super().__init__()
+        if not Z3_AVAIL:
+            raise ImportError('z3-solver is required to use CrosstalkAdaptiveSchedule')
+        self.backend_prop = backend_prop
+        self.crosstalk_prop = crosstalk_prop
+        self.gate_id = {}
+        self.bp_u1_err = {}
+        self.bp_u1_dur = {}
+        self.bp_u2_err = {}
+        self.bp_u2_dur = {}
+        self.bp_u3_err = {}
+        self.bp_u3_dur = {}
+        self.bp_cx_err = {}
+        self.bp_cx_dur = {}
+        self.bp_t1_time = {}
+        self.bp_t2_time = {}
+
+        # Z3 variables
+        self.gate_start_time = {}
+        self.gate_duration = {}
+        self.gate_fidelity = {}
+        self.overlap_amounts = {}
+        self.overlap_indicator = {}
+        self.qubit_lifetime = {}
+        self.dag_overlap_set = {}
+        self.xtalk_overlap_set = {}
+        self.opt = Optimize()
+        self.measured_qubits = []
+        self.measure_start = None
+        self.last_gate_on_qubit = None
+        self.first_gate_on_qubit = None
+        self.fidelity_terms = []
+        self.coherence_terms = []
+        self.model = None
+        self.weight_factor = weight_factor
+        if measured_qubits is None:
+            self.input_measured_qubits = []
+        else:
+            self.input_measured_qubits = measured_qubits
+        self.dag = None
+
+    def powerset(self, iterable):
+        """
+        Finds the set of all subsets of the given iterable
+        This function is used to generate constraints for the Z3 optimization
+        """
+        l_s = list(iterable)
+        return chain.from_iterable(combinations(l_s, r) for r in range(len(l_s)+1))
+
+    def parse_backend_properties(self):
+        """
+        This function assumes that T1, T2 times are in microseconds and
+        gate times are in nanoseconds in backend.properties()
+        """
+        backend_prop = self.backend_prop
+        qid = 0
+        for qinfo in backend_prop.qubits:
```

Comment:
Why not use these methods instead?
```
for qid in range(n_qubits):  # can get num qubits from config
    self.bp_t1_time[qid] = backend_prop.t1(qid)  # * 1e9 if you want it to be in nanoseconds
    self.bp_t2_time[qid] = backend_prop.t2(qid)

self.bp_u1_error = round(backend_prop.gate_error('u1', qid), NUM_PREC)
self.bp_u1_dur = backend_prop.gate_length('u1', qid)
```
etc. Also, I have seen `CX_01` treated separately from `CX_10`. They may have different error rates. It seems like it would be less code to remove the assumption that they are the same (maybe you have another reason for making them the same).
Rather than filling in these attributes, I'd recommend just using the backend methods as you need them

Reply:
I've updated the function to use these methods. Please take a look.

Why do we ignore directions?
1. CX_01 and CX1_0 have error rates which are very close (differing by a few 1q ops). Error rates because of crosstalk are significantly higher. From the optimization perspective, gate direction differences don't really affect the objective significantly.
2. The optimization requires crosstalk characterization data. Adding in CX directions will increase characterization overhead by 4X which is a waste of machine cycles. e.g., for crosstalk charaterization on CX_01 and CX_23, we will have to run simulatneous RB on (CX_01, CX_23), (CX_10, CX_23), (CX_01, CX_32), (CX_10, CX_32).